doc/cascadia/Keybindings-Arguments.md

```diff
@@ -0,0 +1,315 @@
+the `args` object.
+
+Also, pay attention to the last set of keybindings, the `splitProfile` ones.
+This is a function that requires two arguments, both a `orientation` and a
+`profileIndex`. Before this change we would have needed to create 20 separate
+actions (10 profile indicies * 2 directions) to handle these cases. Now it can
+be done with a single action that can be much more flexible in its
+implementation.
+
+### Parsing KeyBinding Arguments
+
+We'll add a new `IKeyBindingArgs` interface. All current keybinding events will
+be changed from their current types to `TypedEventHandler`s. These
+`TypedEventHandler`s second param will always be an instance of
+`IKeyBindingArgs`. So for example:
+
+```csharp
+
+delegate void CopyTextEventArgs();
+delegate void NewTabEventArgs();
+delegate void NewTabWithProfileEventArgs(Int32 profileIndex);
+// ...
+
+[default_interface]
+runtimeclass AppKeyBindings : Microsoft.Terminal.Settings.IKeyBindings
+{
+    event CopyTextEventArgs CopyText;
+    event NewTabEventArgs NewTab;
+    event NewTabWithProfileEventArgs NewTabWithProfile;
+```
+
+Becomes:
+
+```csharp
+interface IKeyBindingArgs {
+    Boolean Handled;
+}
+
+runtimeclass CopyTextEventArgs : IKeyBindingArgs
+{
+    Boolean CopyWhitespace;
+}
+runtimeclass NewTabEventArgs : IKeyBindingArgs { }
+runtimeclass NewTabWithProfileEventArgs : IKeyBindingArgs
+{
+    Int32 ProfileIndex;
+}
+
+[default_interface]
+runtimeclass AppKeyBindings : Microsoft.Terminal.Settings.IKeyBindings
+{
+    event Windows.Foundation.TypedEventHandler<AppKeyBindings, CopyTextEventArgs> CopyText;
+    event Windows.Foundation.TypedEventHandler<AppKeyBindings, NewTabEventArgs> NewTab;
+    event Windows.Foundation.TypedEventHandler<AppKeyBindings, NewTabWithProfileEventArgs> NewTabWithProfile;
+```
+
+We'll also change our existing map in the `AppKeyBindings` implementation.
+Currently, it's a `std::unordered_map<KeyChord, ShortcutAction, ...>`, which
+uses the `KeyChord` to lookup the `ShortcutAction`. We'll need to introduce a
+new type `ActionAndArgs`:
+
+```csharp
+runtimeclass ActionAndArgs
+{
+    ShortcutAction action;
+    IKeyBindingArgs args;
+}
+```
+
+and we'll change the map in `AppKeyBindings` to a `std::unordered_map<KeyChord,
+ActionAndArgs, ...>`.
+
+When we're parsing keybindings, we'll need to construct args for each of the
+events to go with each binding. When we find some key chord bound to a given
+Action, we'll construct the `IKeyBindingArgs` for that action. For many actions,
+these args will be an empty class. However, when we do find an action that needs
+additional parsing, `AppKeyBindingsSerialization` will do the extra work to
+parse the args for that action.
+
+Once the `IKeyBindingArgs` is built for the keybinding, we'll set it in
+`AppKeyBindings` with a updated `AppKeyBindings::SetKeyBinding` call.
+`SetKeyBinding`'s signature will be updated to take a `ActionAndArgs` instead.

```

Comment:
If the action cannot exist without args (most can't, it seems?) do we need a different abstraction?